### PR TITLE
Improve performance for working with indices

### PR DESCRIPTION
**Pull request by fpetersen-gl**
## Description

Don't iterate over all IndexSets for a given indexName, instead use `getForIndex(String)` directly to avoid unnecessary calls. Also converted some loops to streaming for further cleanup of the code.

## Motivation and Context

Fixes issue #18563.

## How Has This Been Tested?

There's an new test to simulate having 500 IndexSets and checking if only the one responsible for a given specific index is being called.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by patrickmann** [comment]
Needs a changelog or [/nocl](https://graylogdocumentation.atlassian.net/wiki/spaces/GLE/pages/2518908931/Changelog+Entries#CI-Check) directive.
Since the PR doesn't really change functionality or user-visible behavior, I would go with `/nocl`.

**Comment by fpetersen-gl** [comment]
I see, thanks for the pointer!
Just to confirm: "user-visible behavior" means only changes in the UI? Because at least the reporter of the bug should see (or rather feel) a difference.

**Review by patrickmann — CHANGES_REQUESTED** [review]

File: graylog2-server/src/test/java/org/graylog2/indexer/MongoIndexSetRegistryTest.java [hunk 1]
```diff
@@ -266,4 +273,95 @@ public void isManagedIndexWithUnmanagedIndexReturnsFalse() {
...
+
+    @Test
+    public void isCurrentWriteIndexSingleIndexSetConfigHappyPath() {
+        final String idxName = "index";
+        final IndexSetConfig indexSetConfig = mock(IndexSetConfig.class);
+        final List<IndexSetConfig> indexSetConfigs = Collections.singletonList(indexSetConfig);
+        final MongoIndexSet indexSet = mock(MongoIndexSet.class);
+        when(mongoIndexSetFactory.create(indexSetConfig)).thenReturn(indexSet);
+        when(indexSetService.findAll()).thenReturn(indexSetConfigs);
+        //End caching
```
[the remark this answers is not in the document]
Comment (fpetersen-gl) [reply graylog2-server/src/test/java/org/graylog2/indexer/MongoIndexSetRegistryTest.java:285]: The mock-definitions before are used for caching of [`findAllMongoIndexSets`](https://github.com/Graylog2/graylog2-server/blob/master/graylog2-server/src/main/java/org/graylog2/indexer/MongoIndexSetRegistry.java#L96-L104). But you're right, it's not easy to understand.

File: graylog2-server/src/main/java/org/graylog2/indexer/MongoIndexSetRegistry.java [hunk 2]
```diff
@@ -223,24 +204,15 @@ public boolean isUp() {
...
-        for (MongoIndexSet indexSet : findAllMongoIndexSets()) {
-            if (indexSet.getActiveWriteIndex() != null && indexSet.getActiveWriteIndex().equals(indexName)) {
-                return true;
-            }
-        }
-
-        return false;
+        return getForIndex(indexName)
+                .map(indexSet -> Objects.equals(indexSet.getActiveWriteIndex(), indexName))
+                .orElse(false);
```
[the remark this answers is not in the document]
Comment (patrickmann) [reply graylog2-server/src/main/java/org/graylog2/indexer/MongoIndexSetRegistry.java:215]: You're right - the bottleneck are the calls to ES/OS, not the Mongo operations

File: graylog2-server/src/test/java/org/graylog2/indexer/MongoIndexSetRegistryTest.java [hunk 3]
```diff
@@ -266,4 +274,93 @@ public void isManagedIndexWithUnmanagedIndexReturnsFalse() {
...
+
+        verify(indexSet).isManagedIndex(idxName);
+        verify(indexSetService).findAll();
+        verifyNoMoreInteractions(indexSetConfig, indexSet);
+    }
+
+    @Test
+    public void isCurrentWriteIndex500IndexSetConfigNoneFoundDueToMismatchingIndexName() {
+        final String idxName = "index";
+        final int noOfIndices = 500;
```
Comment (patrickmann) [inline graylog2-server/src/test/java/org/graylog2/indexer/MongoIndexSetRegistryTest.java:314]: Why 500? Seems like any number >1 would be pretty much equivalent for testing.
Test runtime is a big driver of our overall build time, so we want to keep that as low as possible (without sacrificing coverage).

Comment (fpetersen-gl) [reply graylog2-server/src/test/java/org/graylog2/indexer/MongoIndexSetRegistryTest.java:314]: The number is from the original issue. But you're right, the problem can also be tested with only two indices.

**Review by patrickmann — APPROVED** [review]
Nice first issue - it's my pleasure to merge your first PR

**Comment by HenryTheSir** [comment]
Hi @bernd , 

can we get a backport of this fix to 6.1?

Thanks!

**Pull request merged**